File: my-code/283/main.go
package main

// @Time     : 2021/5/30 21:55
// @Author   : Ranshi
// @File     : 283. 移动零.go

// 冒泡思想
func _moveZeroes(nums []int) {
	for right := len(nums) - 1; right >= 0; right-- {
		for idx := 0; idx < right; idx++ {
			if nums[idx] == 0 && nums[idx+1] != 0 {
				nums[idx], nums[idx+1] = nums[idx+1], nums[idx]
			}
		}
		if nums[right] != 0 {
			break
		}
	}
}

// 双指针
func moveZeroes(nums []int) {
	zeros, length := 0, len(nums)
	for idx := 0; idx < length; idx++ {
		if nums[idx] == 0 {
			zeros++
		} else {
			nums[idx-zeros] = nums[idx]
		}
	}
	for i := length - 1; i >= length-zeros; i-- {
		nums[i] = 0
	}
}

File: my-code/283/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/07/27 19:28
# @Author   : Ranshi
# @File     : 123.py

from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        zero_num = 0
        for idx in range(len(nums)):
            if nums[idx] == 0:
                zero_num += 1
            else:
                nums[idx - zero_num] = nums[idx]
        for idx in range(-1, -zero_num - 1, -1):
            nums[idx] = 0


if __name__ == "__main__":
    nums = [0, 1, 0, 3, 12]
    Solution().moveZeroes(nums)
    print(nums)
